### PR TITLE
Npc attack

### PR DESCRIPTION
**Pull request by KorGgenT**
#### Summary
Features "Implement new logic for NPCs to choose their attacks"

#### Purpose of change
We've been wanting to improve the npc ai, but afaik noone wanted to carry that torch. so here i am! :D

#### Describe the solution
requires #47207 as it has commits from this PR to break it up a bit.

implement a new `npc_attack` class with child classes for the different styles of attacks. in this PR, that is `melee`, `gun`, `throw`, and `activate_item`. throw has a fairly basic implementation because i needed the npc to understand that grenades needed to be thrown, and the activate item attack is for feature parity with npcs using grenades. Unfortunately, it looks pretty involved to properly estimate when it's a good idea to throw a grenade, so the likelyhood of an npc throwing a grenade is "not unless i'm surrounded, can't run, and there's no other option". grenades will have to be their own PR.

add a new method in npc to evaluate all of its possible attack options and save the `npc_attack` and its data (called `npc_attack_effectiveness`) into the ai_cache. then just activate it via npc_attack::use.

#### Describe alternatives you've considered
Watching npcs throw their pants at zombies

#### Testing
i'm writing some tests, and additionally i have run some live tests where i equipped an npc with a specific item and watched what they did with it. i can confirm that they throw items, use attack spells, fire guns, and use melee weapons. they also use grenades if their surrounded and are naked! (if they don't choose to run away first)
I'm drafting this to get it out in the open for some code review, though i'm pretty aware that i'm not done writing up the tests.

#### Additional context
follow ups should be, in no particular order:
- spell
- bionics
- mutations
- grenades
- some other item activations perhaps?

### CODE REVIEW

File: src/npc_attack.h [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+         *  target - the desired target of the npc. valued at 100%
+         */
+        virtual npc_attack_effectiveness evaluate( const npc &source, const Creature *target ) const = 0;
+        virtual void use( npc &source, const tripoint &location ) const = 0;
+        /**
+         *  For debug information. returns all evaluated effectivenesses.
+         *  This is abstracted out in evaluate() as it's faster to not use a container
+         */
+        virtual std::vector<npc_attack_effectiveness> all_evaluations( const npc &source,
+                const Creature *target ) const = 0;
```
Comment (BrettDong) [inline src/npc_attack.h:55]:
```suggestion
                const Creature *target ) const = 0;
                
                virtual ~npc_attack() = default;
```

File: src/npc_attack.h [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+};
```
Comment (BrettDong) [inline src/npc_attack.h:56]:
```suggestion
        virtual std::vector<npc_attack_effectiveness> all_evaluations( const npc &source,
                const Creature *target ) const = 0;
        
        virtual ~npc_attack() = default;
};
```

File: src/npcmove.cpp [hunk 3]
```diff
@@ -371,6 +373,20 @@ static bool too_close( const tripoint &critter_pos, const tripoint &ally_pos, co
...
+{
+    int distance = INT_MAX;
+    for( const weak_ptr_fast<Creature> buddy : friends ) {
+        for( const weak_ptr_fast<Creature> enemy : hostile_guys ) {
+            if( buddy.expired() || enemy.expired() ) {
+                continue;
+            }
+            distance = std::min( distance, rl_dist( buddy.lock()->pos(), enemy.lock()->pos() ) );
+        }
+    }
```
Comment (BrettDong) [inline src/npcmove.cpp:386]:
```suggestion
    for( const weak_ptr_fast<Creature> buddy : friends ) {
        if( !buddy.expired() ) {
            for( const weak_ptr_fast<Creature> enemy : hostile_guys ) {
                if( !enemy.expired()) {
                    distance = std::min( distance, rl_dist( buddy.lock()->pos(), enemy.lock()->pos() ) );
                }
            }
        }
    }
```

**Comment by actual-nh** [comment]
I have been contemplating how to have a monster not use a particular option unless pressed - something more than "only do it every X turns". Possibly dependent on Anger above starting and Morale below starting?

**Comment by BrettDong** [comment]
Need to add virtual destructor to other classes as well.

**Comment by KorGgenT** [comment]
i really don't understand why exactly

**Comment by KorGgenT** [comment]
blergh, i've run both the astyle and json linting tools from VS but they didn't do it

File: src/npc_attack.cpp [hunk 4]
```diff
@@ -0,0 +1,652 @@
...
+npc_attack_effectiveness npc_attack_gun::evaluate(
+    const npc &source, const Creature *target ) const
+{
+    npc_attack_effectiveness effectiveness;
+    if( !can_use( source ) ) {
+        return effectiveness;
+    }
+    const int time_penalty = base_time_penalty( source );
+    for( const tripoint &targetable_point : targetable_points( source ) ) {
+        if( Creature *critter = g->critter_at( targetable_point ) ) {
```
Comment (BrettDong) [inline src/npc_attack.cpp:383]:
```suggestion
        if( g->critter_at( targetable_point ) != nullptr ) {
```

File: src/npc_attack.cpp [hunk 5]
```diff
@@ -0,0 +1,652 @@
...
+std::vector<npc_attack_effectiveness> npc_attack_gun::all_evaluations( const npc &source,
+        const Creature *target ) const
+{
+    std::vector<npc_attack_effectiveness> effectiveness;
+    if( !can_use( source ) ) {
+        return effectiveness;
+    }
+    const int time_penalty = base_time_penalty( source );
+    for( const tripoint &targetable_point : targetable_points( source ) ) {
+        if( Creature *critter = g->critter_at( targetable_point ) ) {
```
Comment (BrettDong) [inline src/npc_attack.cpp:404]:
```suggestion
        if( g->critter_at( targetable_point ) != nullptr ) {
```

File: src/npc_attack.h [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+/**
+ *  The type of npc_attack that is a spell.
+ *  Assumption: if we made it this far, the npc is able to cast the spell.
+ */
+class npc_attack_spell : public npc_attack
+{
+        const spell_id attack_spell_id;
+    public:
+        npc_attack_spell( const spell_id &attack_spell_id ) : attack_spell_id( attack_spell_id ) {}
```
Comment (BrettDong) [inline src/npc_attack.h:68]:
```suggestion
        explicit npc_attack_spell( const spell_id &attack_spell_id ) : attack_spell_id( attack_spell_id ) {}
```

File: src/npc_attack.h [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        int base_time_penalty( const npc &source ) const;
+        npc_attack_effectiveness evaluate_tripoint(
+            const npc &source, const Creature *target, const tripoint &location ) const;
+};
+
+class npc_attack_melee : public npc_attack
+{
+        item &weapon;
+    public:
+        npc_attack_melee( item &weapon ) : weapon( weapon ) {}
```
Comment (BrettDong) [inline src/npc_attack.h:84]:
```suggestion
        explicit npc_attack_melee( item &weapon ) : weapon( weapon ) {}
```

File: src/npc_attack.h [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+        int base_time_penalty( const npc &source ) const;
+        npc_attack_effectiveness evaluate_critter( const npc &source, const Creature *target,
+                Creature *critter ) const;
+};
+
+class npc_attack_gun : public npc_attack
+{
+        const gun_mode gunmode;
+    public:
+        npc_attack_gun( const gun_mode &gunmode ) : gunmode( gunmode ) {}
```
Comment (BrettDong) [inline src/npc_attack.h:101]:
```suggestion
        explicit npc_attack_gun( const gun_mode &gunmode ) : gunmode( gunmode ) {}
```

File: src/npc_attack.h [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        int base_time_penalty( const npc &source ) const;
+        npc_attack_effectiveness evaluate_tripoint(
+            const npc &source, const Creature *target, const tripoint &location ) const;
+};
+
+class npc_attack_throw : public npc_attack
+{
+        item &thrown_item;
+    public:
+        npc_attack_throw( item &thrown_item ) : thrown_item( thrown_item ) {}
```
Comment (BrettDong) [inline src/npc_attack.h:118]:
```suggestion
        explicit npc_attack_throw( item &thrown_item ) : thrown_item( thrown_item ) {}
```

File: src/npc_attack.h [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+        int base_penalty( const npc &source ) const;
+        npc_attack_effectiveness evaluate_tripoint(
+            const npc &source, const Creature *target, const tripoint &location ) const;
+};
+
+class npc_attack_activate_item : public npc_attack
+{
+        item &activatable_item;
+    public:
+        npc_attack_activate_item( item &activatable_item ) : activatable_item( activatable_item ) {}
```
Comment (BrettDong) [inline src/npc_attack.h:135]:
```suggestion
        explicit npc_attack_activate_item( item &activatable_item ) : activatable_item( activatable_item ) {}
```

File: src/character.cpp [hunk 11]
```diff
@@ -2655,7 +2655,7 @@ int Character::get_mod_stat_from_bionic( const character_stat &Stat ) const
     return ret;
 }
 
-int Character::get_standard_stamina_cost( item *thrown_item )
+int Character::get_standard_stamina_cost( const item *thrown_item ) const
```
Comment (mlangsdorf) [inline src/character.cpp:2658]: These changes should pulled out into a separate cleanup PR.

File: src/item.cpp [hunk 12]
```diff
@@ -1484,7 +1484,7 @@ static const double hits_by_accuracy[41] = {
     9993, 9997, 9998, 9999, 10000 // 16 to 20
 };
 
-double item::effective_dps( const Character &guy, monster &mon ) const
```
Comment (mlangsdorf) [inline src/item.cpp:1487]: These changes should also be a separate PR; they're correct whether or not we do the NPC AI rework.

File: src/item.cpp [hunk 13]
```diff
@@ -1519,19 +1519,19 @@ double item::effective_dps( const Character &guy, monster &mon ) const
     // sum average damage past armor and return the number of moves required to achieve
     // that damage
     const auto calc_effective_damage = [ &, moves_per_attack]( const double num_strikes,
-    const bool crit, const Character & guy, monster & mon ) {
-        monster temp_mon = mon;
+    const bool crit, const Character & guy, Creature & mon ) {
+        Creature *temp_mon = &mon;
```
Comment (mlangsdorf) [inline src/item.cpp:1523]: Why did this become a pointer instead of a reference?  I can't see a reason, but I'm probably missing something.  Could you add a comment in the code explaining why?

File: src/magic.cpp [hunk 14]
```diff
@@ -565,6 +569,18 @@ int spell::min_leveled_damage() const
     return type->min_damage + std::round( get_level() * type->damage_increment );
 }
 
+float spell::dps( const Character &caster, const Creature & ) const
```
Comment (mlangsdorf) [inline src/magic.cpp:572]: another set of changes that could be a separate PR.

File: src/npc_attack.cpp [hunk 15]
```diff
@@ -0,0 +1,652 @@
...
+    const map &here = get_map();
+    for( const tripoint &line_point : line ) {
+        if( here.impassable( line_point ) ) {
+            return true;
+        }
+    }
+    return false;
+}
+
+bool npc_attack_effectiveness::operator>( const npc_attack_effectiveness &rhs ) const
```
Comment (mlangsdorf) [inline src/npc_attack.cpp:51]: can we call this structure `npc_attack_rating` and save 7 characters per use? Not a big deal, just a thought.

**Review by mlangsdorf — COMMENTED** [review]
Overall, it looks good and I didn't see any structural problems in the code. @Aivean should probably look over it for performance improvements but I'll leave that to his expertise.

Submission wise, I think a bunch of this can be broken into separate PRs for easier review. There's a lot of generic infrastructure work before you get into actually implementing the changes to the NPC AI. Normally, I wouldn't make such an issue about it, but this is such a large PR that it usefully could be trimmed.

File: src/character.cpp (hunk 11 above)
Comment (KorGgenT) [reply src/character.cpp]: this is a single line of change, i missed it in my other PR that pulled out the plumbing bits, but i'll leave it in this PR unless we're scraping to make it smaller

File: src/item.cpp (hunk 13 above)
Comment (KorGgenT) [reply src/item.cpp]: this was a change in my other PR and i missed the window to edit it before it got merged. this PR doesn't touch that file atm, but i can make sure to note down to go back and re-evaluate why i did that.

**Comment by KorGgenT** [comment]
please excuse the individual commits for now, i'm trying to do miniscule changes to fix up the error messages in order to squash them into the correct commit later.

**Comment by KorGgenT** [comment]
unless i'm reading that general build matrix wrong, looks like it's nearly ready. i'm just going to rebase it and squash my commits into the 4 i had before i undraft it. i might write a few more tests but this thing's been stewing for a long time.

**Comment by KorGgenT** [comment]
all right it should be ready now.

File: src/npc_attack.cpp [hunk 16]
```diff
@@ -0,0 +1,518 @@
...
+        source.do_reload( weapon );
+        add_msg_debug( "%s is reloading %s", source.disp_name(), weapon.display_name() );
+        return;
+    }
+
+    const int dist = rl_dist( source.pos(), location );
+
+    if( source.aim_per_move( weapon, source.recoil ) > 0 &&
+        source.confident_shoot_range( weapon, source.get_most_accurate_sight( weapon ) ) >= dist ) {
+        add_msg_debug( "%s is aiming" );
```
Comment (Night-Pryanik) [inline src/npc_attack.cpp:200]:
```suggestion
        add_msg_debug( "%s is aiming", source.disp_name() );
```

File: src/npc_attack.cpp [hunk 17]
```diff
@@ -0,0 +1,518 @@
...
+    const int damage = source.thrown_item_total_damage_raw( single_item );
+    float dps = damage / throw_mult;
+    const int distance_to_me = rl_dist( location, source.pos() );
+    const int distance_penalty = std::max( distance_to_me - 1,
+                                           ( source.closest_enemy_to_friendly_distance() - 1 ) * 2 );
+    if( friendly_fire ) {
+        // hitting a neutral creature isn't exactly desired, but it's a lot less than a friendly.
+        // if friendly fire is on, we don't care too much, though if an available hit doesn't damage them it would be better.
+        attitude_mult = npc_attack_constants::attitude_multiplier.at( Creature::Attitude::NEUTRAL );
+    }
```
Comment (Night-Pryanik) [inline src/npc_attack.cpp:504]: Looks like a duplicate of code on lines 488-492?

File: tests/npc_attack_test.cpp [hunk 18]
```diff
@@ -0,0 +1,285 @@
...
+    CHECK( *wep_effectiveness.value() == data.base_effectiveness );
+    zomble->set_hp( 1 );
+    wep_effectiveness = wep.evaluate( main_npc, zomble );
+    // effectiveness skyrockets when the zombie will die in one hit
+    CHECK( *wep_effectiveness.value() == data.deathblow );
+
+    npc_attack_setup::spawn_zombie_at_range( 4 );
+    monster *dist_target = g->critter_at<monster>( main_npc_start_tripoint + tripoint( 4, 0, 0 ) );
+    wep_effectiveness = wep.evaluate( main_npc, dist_target );
+    // even if the distant target is the npc's ctual target,
```
Comment (Night-Pryanik) [inline tests/npc_attack_test.cpp:91]:
```suggestion
    // even if the distant target is the npc's actual target,
```

File: data/mods/TEST_DATA/items.json [hunk 19]
```diff
@@ -1044,6 +1044,41 @@
...
+    "name": { "str": "TEST bullwhip" },
+    "description": "A long strip of braided leather with a handle on one end.  Originally developed to settle disagreements with cattle, it's better used now for flaying the rotten flesh off of walking corpses.  Great for when a problem comes along.",
+    "weight": "3460 g",
+    "volume": "2 L",
+    "price": 3500,
+    "price_postapoc": 500,
+    "material": [ "leather" ],
+    "flags": [ "REACH_ATTACK", "REACH3", "WHIP" ],
+    "cutting": 3,
+    "category": "weapons"
```
[the remark this answers is not in the document]
Comment (KorGgenT) [reply data/mods/TEST_DATA/items.json:1061]: no i specifically don't want it to use copy-from in case the regular item gets updated. it won't break the test this way.

**Comment by Salty-Panda** [comment]
Will it be applicable to turrets?
As in: small caliber turrets only shoot if can deal damage, .50 cal prioritizes high armor targets.

**Comment by actual-nh** [comment]
> Will it be applicable to turrets?
> As in: small caliber turrets only shoot if can deal damage, .50 cal prioritizes high armor targets.

I like the idea, but IIRC turrets are monsters, not NPCs. (There really should be some/more AI shared between them, though!)

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/magiclysm-a-total-fantasy-overhaul/20612/512
